File: mobile/lib/models/habit.dart
import 'dart:ui';

class Habit {
  final int id;
  final int userId;
  final String name;
  final String description;
  final String category;
  final String frequency;
  final int cooldownDays;
  final int dailyTarget;
  final String? targetTime;
  final String? reminderTime;
  final String color;
  final String icon;
  final bool isActive;
  final DateTime createdAt;
  final int currentStreak;
  final int bestStreak;
  final bool completedToday;
  final int todayCompletions;
  final double completionRate;

  Habit({
    required this.id,
    required this.userId,
    required this.name,
    required this.description,
    required this.category,
    required this.frequency,
    this.cooldownDays = 1,
    this.dailyTarget = 1,
    this.targetTime,
    this.reminderTime,
    required this.color,
    required this.icon,
    required this.isActive,
    required this.createdAt,
    this.currentStreak = 0,
    this.bestStreak = 0,
    this.completedToday = false,
    this.todayCompletions = 0,
    this.completionRate = 0.0,
  });

  factory Habit.fromJson(Map<String, dynamic> json) {
    return Habit(
      id: json['id'],
      userId: json['user_id'],
      name: json['name'],
      description: json['description'] ?? '',
      category: json['category'] ?? 'other',
      frequency: json['frequency'] ?? 'daily',
      cooldownDays: json['cooldown_days'] ?? 1,
      dailyTarget: json['daily_target'] ?? 1,
      targetTime: json['target_time'],
      reminderTime: json['reminder_time'],
      color: json['color'] ?? '#4CAF50',
      icon: json['icon'] ?? 'check_circle',
      isActive: json['is_active'] ?? true,
      createdAt: DateTime.parse(json['created_at']),
      currentStreak: json['current_streak'] ?? 0,
      bestStreak: json['best_streak'] ?? 0,
      completedToday: json['completed_today'] ?? false,
      todayCompletions: json['today_completions'] ?? 0,
      completionRate: (json['completion_rate'] ?? 0.0).toDouble(),
    );
  }

  Map<String, dynamic> toCreateJson() {
    return {
      'name': name,
      'description': description,
      'category': category,
      'frequency': frequency,
      'cooldown_days': cooldownDays,
      'daily_target': dailyTarget,
      'target_time': targetTime,
      'reminder_time': reminderTime,
      'color': color,
      'icon': icon,
    };
  }

  Color get colorValue {
    try {
      return Color(int.parse(color.replaceFirst('#', '0xFF')));
    } catch (_) {
      return const Color(0xFF4CAF50);
    }
  }
}

File: backend/app/schemas/habit.py
from pydantic import BaseModel
from datetime import datetime, date
from app.models.habit import HabitCategory, HabitFrequency


# --- Habit ---
class HabitCreate(BaseModel):
    name: str
    description: str = ""
    category: HabitCategory = HabitCategory.OTHER
    frequency: HabitFrequency = HabitFrequency.DAILY
    cooldown_days: int = 1  # 1=каждый день, 2=через день...
    target_time: str | None = None  # HH:MM — когда выполнять
    reminder_time: str | None = None  # HH:MM — когда push
    color: str = "#4CAF50"
    icon: str = "check_circle"


class HabitUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    category: HabitCategory | None = None
    frequency: HabitFrequency | None = None
    cooldown_days: int | None = None
    target_time: str | None = None
    reminder_time: str | None = None
    color: str | None = None
    icon: str | None = None
    is_active: bool | None = None


class HabitResponse(BaseModel):
    id: int
    user_id: int
    name: str
    description: str
    category: HabitCategory
    frequency: HabitFrequency
    cooldown_days: int = 1
    target_time: str | None
    reminder_time: str | None
    color: str
    icon: str
    is_active: bool
    created_at: datetime
    current_streak: int = 0
    completion_rate: float = 0.0

    model_config = {"from_attributes": True}


# --- Habit Log ---
class HabitLogCreate(BaseModel):
    habit_id: int
    date: date
    completed: bool = True
    note: str | None = None
    skipped_reason: str | None = None


class HabitLogResponse(BaseModel):
    id: int
    habit_id: int
    date: date
    completed: bool
    completed_at: datetime | None
    note: str | None
    skipped_reason: str | None
    created_at: datetime

    model_config = {"from_attributes": True}
